### PR TITLE
Bypass firewall

### PR DESCRIPTION
**Pull request by outscale-jju**

### CODE REVIEW

File: api/server/graph.cc [hunk 1]
```diff
@@ -463,26 +463,40 @@ std::string Graph::nic_add(const app::Nic &nic) {
         return "";
     }
 
-    // Link branch (inside)
-    if (pg_brick_link(gn.firewall.get(),
-                      gn.antispoof.get(), &app::pg_error) < 0) {
-        PG_ERROR_(app::pg_error);
+    // Build branch and set head
+    if (nic.bypass_filtering) {
```
Comment (outscale-mgo) [inline api/server/graph.cc]: you coukd have write:
if (!nic.bypass_filtering) {
...
} else if (app::config.packet_trace) {
...
} else {
...
}

same result with less lines of code

File: api/server/api.cc [hunk 2]
```diff
@@ -135,7 +135,7 @@ bool API::action_nic_add(const app::Nic &nic, std::string *path,
     // Add NIC in model
     std::pair<std::string, app::Nic> p(nic.id, nic);
     app::model.nics.insert(p);
-    return true;
+    return path->size();
```
Comment (outscale-mgo) [inline api/server/api.cc]: the message is not very explicit, "return false on empty path"
In C I would use `return !! path->size()` but as bool is a type in C++, I guess the `!!` is not needed, but maybe more explicit ?

**Comment by outscale-mgo** [comment]
can you rebase this please

File: tests/scenario_36/test.sh [hunk 3]
```diff
@@ -84,7 +84,8 @@ cli 0 0 nic del nic_1
 cli 0 0 nic add --ip 42.0.0.1 --mac 42:42:42:42:42:01 --vni 1337 --id nic_1
 cli 0 0 nic del nic_1 nic_2 nic_not_here
 cli 0 0 nic add --ip 42.0.0.1 --mac 42:42:42:42:42:01 --vni 1337 --id nic_1
-cli 0 0 nic add --ip 42.0.0.2 --mac 42:42:42:42:42:01 --vni 1337 --id nic_2
+cli 0 0 nic add --ip 42.0.0.2 --mac 42:42:42:42:42:01 --vni 1337 --id nic_2 --enable-antispoof
```
Comment (outscale-jju) [inline tests/scenario_36/test.sh]: Just to use the option

File: api/server/graph.cc [hunk 4]
```diff
@@ -473,26 +473,41 @@ std::string Graph::nic_add(const app::Nic &nic) {
...
-        PG_ERROR_(app::pg_error);
+
+    // Build branch and set head
+    if (nic.bypass_filtering) {
+        if (app::config.packet_trace) {
+            gn.head = gn.sniffer;
+            if (pg_brick_link(gn.sniffer.get(), gn.vhost.get(),
+                              &app::pg_error) < 0) {
+                PG_ERROR_(app::pg_error);
+                return false;
```
Comment (outscale-mgo) [inline api/server/graph.cc]: this function return  a string

File: api/server/graph.cc [hunk 5]
```diff
@@ -473,26 +473,41 @@ std::string Graph::nic_add(const app::Nic &nic) {
...
+            }
+        } else {
+            gn.head = gn.vhost;
+        }
+    } else {
+        gn.head = gn.firewall;
+        if (pg_brick_link(gn.firewall.get(),
+                          gn.antispoof.get(), &app::pg_error) < 0) {
+            PG_ERROR_(app::pg_error);
+            return false;
```
Comment (outscale-mgo) [inline api/server/graph.cc]: same here

File: api/server/graph.h [hunk 6]
```diff
@@ -224,6 +224,8 @@ class Graph {
     /* VM branch. */
     struct GraphNic {
        std::string id;
+       // head is a pointer to the first brick in the branch
+       BrickShrPtr head;
```
Comment (outscale-mgo) [inline api/server/graph.h]: can't we use a 'struct pg_brick *' ?

**Pull request merged**